e2e/tests/upgrade-test.rs

```diff
@@ -1,36 +1,438 @@
+
+    // a unique name for the test combination we are running
+    let name = format!(
+        "upgrade-from-{}-{}-{}-pg{}",
+        from_version,
+        match with_data {
+            true => "with-data",
+            false => "no-data",
+        },
+        flavor,
+        pg_version
+    );
+
+    // figure out which image we should start with in the upgrade process
+    // based on alpine vs ha and postgres version
+    let from_image_uri = from_image(&flavor, &pg_version, &from_version);
+    info!("from image {}", from_image_uri);
+    info!("to image {}", to_image_uri);
+
+    // we'll put our db snapshots to compare in this directory
+    let temp_dir = TempDir::new(&name).expect("unable to create working dir");
+    let working_dir = temp_dir.path();
+    if working_dir.exists() {
+        remove_dir_all(&working_dir).expect("failed to remove working dir");
+    }
+    create_dir_all(&working_dir).expect("failed to create working dir");
+    info!("working dir at {}", working_dir.to_str().unwrap());
+
+    // for the database we're upgrading, we need to use two containers and the data_directory
+    // needs to be persistent in this dir. we'll nest it inside the working directory
+    let data_dir = working_dir.join("db");
+    if !data_dir.exists() {
+        create_dir_all(&data_dir).expect("failed to create working dir and data dir");
+    }
+    // The HA image requires that we map chown 777'd parent directory in, while
+    // the alpine image requires that we map the data directory in.
+    let volume_map_path = match flavor {
+        Flavor::Alpine => "/var/lib/postgresql/data",
+        Flavor::HA => "/var/lib/postgresql",
+    };
+    // We need to chmod 777 this directory so that the HA image's initdb can create a subdirectory
+    let permissions = Permissions::from_mode(0o777);
+    set_permissions(&data_dir, permissions.clone())
+        .expect("failed to chmod 0o777 on the data directory");
+    let data_dir = data_dir.to_str().unwrap();
+    info!("data dir at {}", &data_dir);
+
+    // this dir has our sql scripts. we'll mount it into the docker containers and use them via psql
+    let script_dir = concat!(env!("CARGO_MANIFEST_DIR"), "/scripts");
+    info!("script dir at {}", script_dir);
+
+    /**********************************************************************************************/
+    // BASELINE
+    info!("BASELINE");
+    let baseline_blueprint = PostgresContainerBlueprint::new()
+        .with_image_uri(to_image_uri.clone())
+        .with_volume(script_dir, "/scripts")
+        .with_env_var("PGDATA", "/var/lib/postgresql/data")
+        .with_db("db")
+        .with_user("postgres");
+    let baseline_container = baseline_blueprint.run();
+    let mut baseline_client = connect(&baseline_blueprint, &baseline_container);
+
+    // determine the versions of the promscale extension we are supposed to upgrade from and to
+    let (from_version, to_version) = {
```

Comment:
The nested lets are a little terse to read

Reply:
done